### PR TITLE
Retry mechanism for testing service connection

### PR DESCRIPTION
**Pull request by grafnu**

### CODE REVIEW

File: forch/device_report_client.py [hunk 1]
```diff
@@ -90,13 +90,19 @@ def _convert_and_handle(self, mac, progress):
...
         try:
             for progress in session:
+                should_retry = False
                 if self._convert_and_handle(mac, progress):
                     self._logger.warning('Terminal state reached for %s', mac)
             self._logger.error('Progress complete for %s', mac)
         except Exception as e:
             self._logger.error('Progress exception: %s', e)
+        self.disconnect(mac)
+        if should_retry:
```
Comment (henry54809) [inline forch/device_report_client.py:103]: It looks like the session progress can be empty, thus session needs to be restarted. Shouldn't we fix the root cause of why session can be empty coming from DAQ?

**Comment by grafnu** [comment]
The root cause was that the client tried to connect to a not-yet-started
server, so it wasn't that the session was empty, it was that it threw an
exception.  The actual path through the code to a retry was through the
exception handler.  I've never seen the case where session was empty, and
I'm not even sure how that would happen.

**Comment by henry54809** [comment]
What about just let forch wait until the channel is ready? https://grpc.github.io/grpc/python/grpc.html#channel-ready-future

**Comment by grafnu** [comment]
Updated to use channel ready future, PTAL.

**Review by henry54809 — APPROVED** [review]

**Pull request merged**